### PR TITLE
[OPS-3200] Add per repo permissions

### PR DESCRIPTION
**Pull request by ShelRoman**

### CODE REVIEW

File: plugin/plugin.go [hunk 1]
```diff
@@ -31,11 +31,25 @@ func stringInSlice(str string, slice []string) bool {
 
 // New returns a new validator plugin.
 func New(privilegedUsers []string, userPermissionsRaw map[string]string) validator.Plugin {
-	userPermissions := make(map[string][]string)
+	userPermissions := make(map[string]map[string][]string)
```
Comment (bkk-bcd) [inline plugin/plugin.go]: What if we model permissions with structs and https://github.com/go-yaml/yaml/tree/v3 or json? That seems better than hacking in our own serialization...

```
map["trent"] = `encoded yaml or json here`
```

Comment (ShelRoman) [reply plugin/plugin.go]: That's the lack of knowing Go :)
My thought process has been like that:
We have an env variable with text in the config hence we need to parse text.
Is there an easy way to pass yaml or json from the env variable?

File: plugin/plugin.go [hunk 2]
```diff
@@ -55,25 +70,29 @@ func (p *plugin) Validate(ctx context.Context, req *validator.Request) error {
 		// check if user is privilged to promote to any env
 		if stringInSlice(req.Build.Trigger, p.privilegedUsers) {
 			logrus.Debugf(
```
Comment (bkk-bcd) [inline plugin/plugin.go]: Use fields so the log is more machine parseable. We should actually switch to zap but that's ok for now. Ex: https://github.com/getgoing/drone-amazon-ssm/blob/master/plugin/plugin.go#L56

File: plugin/plugin.go [hunk 3]
```diff
@@ -55,25 +70,29 @@ func (p *plugin) Validate(ctx context.Context, req *validator.Request) error {
...
-			if stringInSlice(req.Build.Deploy, allowedEnvs) {
-				logrus.Debugf(
-					"User %s has been authorized to %s to/on %s env according to user level permissions",
-					req.Build.Trigger, req.Build.Event, req.Build.Deploy,
-				)
-				return nil
+
+			for env, repos := range allowedEnvs {
+				if env == req.Build.Deploy && stringInSlice(req.Repo.Name, repos) {
+					logrus.Debugf(
```
Comment (bkk-bcd) [inline plugin/plugin.go]: Use fields so the log is more machine parseable. We should actually switch to zap but that's ok for now. Ex: https://github.com/getgoing/drone-amazon-ssm/blob/master/plugin/plugin.go#L56

File: plugin/plugin.go [hunk 4]
```diff
@@ -55,25 +70,29 @@ func (p *plugin) Validate(ctx context.Context, req *validator.Request) error {
...
+						req.Build.Trigger, req.Build.Event, req.Build.Deploy, req.Repo.Name,
+					)
+					return nil
+				}
 			}
+
 		}
 
-		logrus.Debugf("user %s not allowed to %s to/on %s", req.Build.Trigger, req.Build.Event, req.Build.Deploy)
+		logrus.Debugf("user %s not allowed to %s to/on %s in %s repo", req.Build.Trigger, req.Build.Event, req.Build.Deploy, req.Repo.Name)
```
Comment (bkk-bcd) [inline plugin/plugin.go]: Use fields so the log is more machine parseable. We should actually switch to zap but that's ok for now. Ex: https://github.com/getgoing/drone-amazon-ssm/blob/master/plugin/plugin.go#L56

**Review by bkk-bcd — APPROVED** [review]
Cool, thanks for the logging tweaks and change to a more parseable and readable format (csv)

**Pull request merged**